### PR TITLE
Add support for extension types

### PR DESCRIPTION
**Pull request by srujzs**
Corresponds to the syntax defined in https://github.com/dart-lang/language/blob/main/accepted/future-releases/extension-types/feature-specification.md#syntax.

- [X] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

### CODE REVIEW

File: test/specs/extension_test.dart
Comment (srujzs) [inline test/specs/extension_test.dart:138]: FYI - this is just a duplicate test.

File: pubspec.yaml
Comment (srujzs) [inline pubspec.yaml:22]: I'm not really sure if this is an issue (and it looks like CI isn't happy with it anyways due to the SDK requirement). Should I just keep it as `^2.0.0`?

File: lib/src/emitter.dart
Comment (jakemac53) [inline lib/src/emitter.dart:363]: consider joining with ', ' instead so we don't have a trailing comma?

Comment (srujzs) [reply lib/src/emitter.dart]: It is indeed what it does - it takes in a separator between any two elements. I do add a space here to make the non-formattable code look a bit better, but maybe I shouldn't for consistency and just wait for the new version of dart_style.

**Comment by jakemac53** [comment]
@srujzs fwiw force pushing in GitHub doesn't work well because it destroys all history (I can't tell what you actually changed in that last push). If you can do a merge from master instead in the future that will help me out :).

File: pubspec.yaml
Comment (jakemac53) [reply pubspec.yaml:22]: Yeah just keep it at ^2.0.0 (it will still get the latest that it can, on whatever SDK it is running).

We could update the SDK versions of the the job is running the formatter if that is desirable, but the important thing is that your formatter version matches that of the job we are running, otherwise you will have issues.

File: test/specs/extension_type_test.dart
Comment (jakemac53) [inline test/specs/extension_type_test.dart:162]: I would expect a space between these, the matcher might just be a bit loose though because I thought you did add that, but might as well update the expectation to show it:

```suggestion
        extension type Foo(int bar) implements num, Object { }
```

**Review by jakemac53 — COMMENTED** [review]
LGTM modulo the formatter change

File: pubspec.yaml
Comment (jakemac53) [reply pubspec.yaml:22]: Oh I should have read this comment more fully :).

We should probably increase the minimum to require a version of dart_style that can actually format extension types.

I just chatted with @munificent and it sounds like we can get a release, so we could change this to ^2.3.4.

**Comment by srujzs** [comment]
> fwiw force pushing in GitHub doesn't work well because it destroys all history (I can't tell what you actually changed in that last push).

Ah sorry about that! I've been pushing to avoid unnecessary commits from polluting the branch but messing with the review is way worse.

File: test/specs/extension_type_test.dart
Comment (srujzs) [reply test/specs/extension_type_test.dart:162]: This was the one change I made before the force push. :)

I changed the separator to not have that space to be consistent with the rest of the emitter and to avoid spaces that aren't needed in the output. Dart format should add the space for us when 2.3.4 is released.

File: pubspec.yaml
Comment (srujzs) [reply pubspec.yaml]: Got it. I guess wait for 2.3.4 then before landing this? I'll do some testing with 2.3.3 and updated SDK constraint to see if any of the bots are unhappy first.

**Comment by jakemac53** [comment]
> Ah sorry about that! I've been pushing to avoid unnecessary commits from polluting the branch but messing with the review is way worse.

No worries, really it is just specific to GitHub and how its reviews work. Fwiw when we merge we squash/merge so its always just one commit. So you don't have to worry about branch history really (and it is useful for reviewing so you can see diffs since last review).

File: test/specs/extension_type_test.dart
Comment (jakemac53) [reply test/specs/extension_type_test.dart:162]: Ah ok, SGTM :)

**Review by jakemac53 — APPROVED** [review]

File: pubspec.yaml
Comment (jakemac53) [reply pubspec.yaml]: as discussed offline either way is OK with me :)

**Pull request merged**
